### PR TITLE
[NETBEANS-3604]Supports for Records (JEP 359) in JDK-14"

### PR DESCRIPTION
**Pull request by singh-akhilesh**
Jira ref : https://issues.apache.org/jira/browse/NETBEANS-3604

### CODE REVIEW

File: java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/Utilities.java [hunk 1]
```diff
@@ -557,7 +557,7 @@ private static int findBodyStartImpl(CompilationInfo info, Tree cltree, Compilat
     
     public static int findBodyStart(final CompilationInfo info, final Tree cltree, final CompilationUnitTree cu, final SourcePositions positions, final Document doc) {
         Kind kind = cltree.getKind();
-        if (!TreeUtilities.CLASS_TREE_KINDS.contains(kind) && kind != Kind.METHOD)
+        if (!TreeUtilities.CLASS_TREE_KINDS.contains(kind) && kind != Kind.METHOD && !cltree.getKind().toString().equals("RECORD"))
```
Comment (lahodaj) [inline java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/Utilities.java]: Unless this was tried already, and rejected for some reason, I would recommend to add Tree.Kind.RECORD into TreeUtilities.CLASS_TREE_KINDS. The purpose of that constant is to keep all kinds that map to ClassTree.

File: java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/DetectorTest.java [hunk 2]
```diff
@@ -508,6 +508,62 @@ public void testIncDecReading230408() throws Exception {
         performTest("IncDecReading230408");
     }
 
+    public void testRecord1() throws Exception {
+        //TODO: check records are available, skip if they are not.
```
Comment (lahodaj) [inline java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/DetectorTest.java]: There should be a SourceVersion check, or something along those lines, and the test case should be skipped if the current javac does not support records. Otherwise, the test will fail when running e.g. with nb-javac without records. (Dtto the other occurrence below.)

File: java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/TestBase.java [hunk 3]
```diff
@@ -35,6 +35,7 @@
 import java.util.Collection;
 import java.util.Comparator;
 import java.util.EnumSet;
+import java.util.Iterator;
```
Comment (lahodaj) [inline java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/TestBase.java]: Ideally, unnecessary changes should be undone.

File: java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java [hunk 4]
```diff
@@ -381,6 +385,35 @@ public static Context resolveContext(CompilationInfo controller, Document doc, i
                             el = e;
                         }
                     }
+                    if (el != null && el.getKind() == ElementKind.METHOD) {
+                        for (Element peer : el.getEnclosingElement().getEnclosedElements()) {
+                            if (peer.getKind().name().contains("RECORD_COMPONENT")) {
+                                try {
+                                    Class<?> recordComponent = Class.forName("javax.lang.model.element.RecordComponentElement", true, VariableTree.class.getClassLoader());
+                                    Method getAccessor = recordComponent.getDeclaredMethod("getAccessor");
+                                    Method getRecordComponents = TypeElement.class.getDeclaredMethod("getRecordComponents");
```
Comment (lahodaj) [inline java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java]: I wonder if the reflection should be separated into TreeShims. But this can wait.

File: java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java [hunk 5]
```diff
@@ -986,6 +1020,14 @@ public Void visitTypeParameter(TypeParameterElement e, Boolean highlightName) {
             return null;
         }
         
+        @Override
+        public Void visitUnknown(Element e, Boolean p) {
+            if (TreeShims.isRecordComponent(e)) {
+                return visitVariable((VariableElement) e, p);
+            }
+            return super.visitUnknown(e, p); //To change body of generated methods, choose Tools | Templates.
```
Comment (lahodaj) [inline java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java]: Ideally comments like "//To change..." should be deleted.

File: java/java.editor/test/unit/src/org/netbeans/modules/editor/java/GoToSupportTest.java [hunk 6]
```diff
@@ -1282,4 +1297,223 @@ private static boolean hasPatterns() {
             return false;
         }
     }
+    
+    public void testRecords1() throws Exception {
+//        try {
```
Comment (lahodaj) [inline java/java.editor/test/unit/src/org/netbeans/modules/editor/java/GoToSupportTest.java]: This should be (presumably?) uncomentted. (Dtto the other occurrences below.)

File: java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/SemanticHighlighterBase.java [hunk 7]
```diff
@@ -573,6 +582,27 @@ private void addModifiers(Element decl, Collection<ColoringAttributes> c) {
             return null;
         }
 
+        private boolean isRecordComponent(ElementKind kind) {
+            return "RECORD_COMPONENT".equals(kind.name());
+        }
+
+        private Element toRecordComponent(Element el) {
```
Comment (lahodaj) [inline java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/SemanticHighlighterBase.java]: Can TreeShims.toRecordComponent be used here?

File: java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/DetectorTest.java (hunk 2 above)
Comment (singh-akhilesh) [reply java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/DetectorTest.java]: Comment addressed.

File: java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/TestBase.java (hunk 3 above)
Comment (singh-akhilesh) [reply java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/TestBase.java]: Comment addressed.

File: java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java (hunk 4 above)
Comment (singh-akhilesh) [reply java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java]: It would be good if we wait and fix this at later stage.

File: java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java (hunk 5 above)
Comment (singh-akhilesh) [reply java/java.editor/src/org/netbeans/modules/editor/java/GoToSupport.java]: Comment addressed.

File: java/java.editor/test/unit/src/org/netbeans/modules/editor/java/GoToSupportTest.java (hunk 6 above)
Comment (singh-akhilesh) [reply java/java.editor/test/unit/src/org/netbeans/modules/editor/java/GoToSupportTest.java]: Comment addressed.

File: java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/SemanticHighlighterBase.java (hunk 7 above)
Comment (singh-akhilesh) [reply java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/SemanticHighlighterBase.java]: Comment addressed.

File: java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/Utilities.java (hunk 1 above)
Comment (singh-akhilesh) [reply java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/Utilities.java]: Tree.Kind.RECORD is a final Enumset, So adding RECORD here as a preview feature would result a compilation failure for the versions below 14.

File: java/java.navigation/src/org/netbeans/modules/java/navigation/ElementScanningTask.java [hunk 8]
```diff
@@ -358,7 +360,7 @@ private Description element2description(final Element e, final Element parent,
         } else if( e instanceof ExecutableElement ) {
             d.htmlHeader = createHtmlHeader(info,  (ExecutableElement)e, info.getElements().isDeprecated(e),d.isInherited, fqn, overridenFrom);
         } else if( e instanceof VariableElement ) {
-            if( !(e.getKind() == ElementKind.FIELD || e.getKind() == ElementKind.ENUM_CONSTANT) )
+            if( !(e.getKind() == ElementKind.FIELD || e.getKind() == ElementKind.ENUM_CONSTANT || e.getKind().name().equals("STATE_COMPONENT")) )
```
Comment (mbien) [inline java/java.navigation/src/org/netbeans/modules/java/navigation/ElementScanningTask.java:363]: @singh-akhilesh should this be `RECORD_COMPONENT`? ElementKind does not have a `STATE_COMPONENT`.

Comment (singh-akhilesh) [reply java/java.navigation/src/org/netbeans/modules/java/navigation/ElementScanningTask.java:363]: Hi @mbien STATE_COMPONENT was part of the amber project and  replace by RECORD_COMPONENT in the java14 https://hg.openjdk.java.net/amber/amber/rev/2a5e841bd91b. Will take necessary action.

Comment (singh-akhilesh) [reply java/java.navigation/src/org/netbeans/modules/java/navigation/ElementScanningTask.java:363]: https://github.com/apache/netbeans/pull/4293
